src/sage/rings/polynomial/multi_polynomial_libsingular.pyx

```diff
@@ -4485,31 +4565,30 @@ cdef class MPolynomial_libsingular(MPolynomial_libsingular_base):
             except Exception:
                 raise NotImplementedError("Factorization of multivariate polynomials over %s is not implemented."%self._parent._base)
 
-        if n_GetChar(_ring.cf) > 1<<29:
-            raise NotImplementedError("Factorization of multivariate polynomials over prime fields with characteristic > 2^29 is not implemented.")
-
-        # I make a temporary copy of the poly in self because singclap_factorize appears to modify it's parameter
-        ptemp = p_Copy(self._poly, _ring)
         iv = NULL
-        sig_on()
         if _ring != currRing:
             rChangeCurrRing(_ring)   # singclap_factorize
-        I = singclap_factorize(ptemp, &iv, 0, _ring)
-        sig_off()
-
-        ivv = iv.ivGetVec()
-        v = [(new_MP(parent, p_Copy(I.m[i], _ring)), ivv[i])
-             for i in range(1, I.ncols)]
-        v = [(f, m) for f, m in v if f != 0]  # we might have zero in there
-        unit = new_MP(parent, p_Copy(I.m[0], _ring))
-
-        F = Factorization(v, unit)
-        F.sort()
-
-        del iv
-        id_Delete(&I, _ring)
-
-        return F
+        start_catch_error()
+        try:
+            try:
+                sig_on()
+                I = singclap_factorize(p_Copy(self._poly, _ring), &iv, 0, _ring)
+                sig_off()
+            finally:
+                if check_error():
+                    raise NotImplementedError("Factorization of multivariate polynomials over prime fields with characteristic > 2^29 is not implemented.")
+
+            ivv = iv.ivGetVec()
+            v = [(f, ivv[i]) for i in range(1, I.ncols)
+                 if (f := new_MP(parent, p_Copy(I.m[i], _ring)))]
```

Comment:
```suggestion
                 if (f := new_MP(parent, p_Copy(I.m[i], _ring))) != 0]
```
I don't really understand why it makes a difference, but apparently it does.

Reply:
Let's see, `new_MP` returns `MPolynomial_libsingular`. So…